lib/SPIRV/SPIRVWriter.cpp

```diff
@@ -414,16 +414,40 @@ SPIRVType *LLVMToSPIRV::transType(Type *T) {
     if (Name == getSPIRVTypeName(kSPIRVTypeName::ConstantPipeStorage))
       return transType(getPipeStorageType(M));
 
-    auto *Struct = BM->openStructType(T->getStructNumElements(), Name.str());
+    const size_t MaxNumElements = 65535 - 2;
+    const size_t NumElements = T->getStructNumElements();
+    size_t SPIRVStructNumElements = NumElements;
+    if (NumElements > MaxNumElements &&
+        BM->isAllowedToUseExtension(
+            ExtensionID::SPV_INTEL_long_constant_composite)) {
+      SPIRVStructNumElements = MaxNumElements;
+    }
+
+    auto *Struct = BM->openStructType(SPIRVStructNumElements, Name.str());
     mapType(T, Struct);
 
+    if (NumElements > MaxNumElements &&
+        BM->isAllowedToUseExtension(
+            ExtensionID::SPV_INTEL_long_constant_composite)) {
+      uint64_t NumOfContinuedInstructions = NumElements / MaxNumElements - 1;
+      for (uint64_t J = 0; J < NumOfContinuedInstructions; J++) {
+        auto *Continued = BM->addTypeStructContinuedINTEL(MaxNumElements);
+        Struct->addContinuedInstruction(
+            static_cast<SPIRVTypeStruct::ContinuedInstType>(Continued));
+      }
+      uint64_t Remains = NumElements % MaxNumElements;
+      if (Remains) {
+        auto *Continued = BM->addTypeStructContinuedINTEL(Remains);
+        Struct->addContinuedInstruction(
+            static_cast<SPIRVTypeStruct::ContinuedInstType>(Continued));
+      }
+    }
+
     SmallVector<unsigned, 4> ForwardRefs;
 
     for (unsigned I = 0, E = T->getStructNumElements(); I != E; ++I) {
       auto *ElemTy = ST->getElementType(I);
-      if ((isa<StructType>(ElemTy) || isa<ArrayType>(ElemTy) ||
-           isa<VectorType>(ElemTy) || isa<PointerType>(ElemTy)) &&
-          recursiveType(ST, ElemTy))
+      if (recursiveType(ST, ElemTy))
```

Comment:
Why this `if` was updated?

Reply:
I was looking through the code and tried to simplify it, so I could implement `OpTypeStructContinuedINTEL` as much as possible close to `OpConstantCompositeContinuedINTEL`, but I turned out I couldn't do it in case recursive type is used. However I left this simplification. I can revert it if you prefer.

Reply:
>  I can revert it if you prefer.

It just seems to be unrelated to the patch, plus it seems to introduce more calls to `recursiveType` function than it used to be (it will be called for each struct member instead of being called only for composite ones)

Reply:
Reverted.